Error in `simplify_network`

### Checklist

- [X] I am using the current [`master`](https://github.com/PyPSA/pypsa-usa/tree/master) branch
- [X] I am running on an up-to-date [`pypsa-usa` environment](https://github.com/PyPSA/pypsa-usa/blob/master/workflow/envs/environment.yaml). Update via `conda env update -f envs/environment.yaml`

### The Issue

When running with `eia` demand data, an error is thrown in simplify network. 

### Steps To Reproduce

1. clone the repo
2. comment out the planning horizon value 
3. run the workflow 

### Expected Behavior

_No response_

### Error Message

```shell
[Sat Feb 24 16:31:17 2024]
rule simplify_network:
    input: data/breakthrough_network/base_grid/western/bus2sub.csv, data/breakthrough_network/base_grid/western/sub.csv, resources/western/elec_base_network_l_pp.nc
    output: resources/western/elec_s.nc
    log: logs/simplify_network/western/elec_s.log
    jobid: 7
    reason: Missing output files: resources/western/elec_s.nc; Input files updated by another job: data/breakthrough_network/base_grid/western/bus2sub.csv, resources/western/elec_base_network_l_pp.nc, data/breakthrough_network/base_grid/western/sub.csv
    wildcards: interconnect=western
    threads: 2
    resources: tmpdir=/tmp, mem_mb=10000, mem_mib=9537

INFO:pypsa.io:Imported network elec_base_network_l_pp.nc has buses, carriers, generators, lines, line_types, links, loads, storage_units, transformers
INFO:__main__:Mapping all network lines onto a single layer
INFO:__main__:Assigning line lengths using haversine function...
INFO:__main__:Aggregating buses to substation level...
Traceback (most recent call last):
  File "/home/trevor/repos/pypsa-usa/workflow/.snakemake/scripts/tmp9zo8mvsl.simplify_network.py", line 199, in <module>
    n = aggregate_to_substations(
        ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/trevor/repos/pypsa-usa/workflow/.snakemake/scripts/tmp9zo8mvsl.simplify_network.py", line 86, in aggregate_to_substations
    clustering = get_clustering_from_busmap(
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/trevor/miniconda3/envs/pypsa-usa/lib/python3.11/site-packages/pypsa/clustering/spatial.py", line 534, in get_clustering_from_busmap
    buses = aggregatebuses(n, busmap, custom_strategies=bus_strategies)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/trevor/miniconda3/envs/pypsa-usa/lib/python3.11/site-packages/pypsa/clustering/spatial.py", line 341, in aggregatebuses
    aggregated = n.buses.groupby(busmap).agg(strategies)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/trevor/miniconda3/envs/pypsa-usa/lib/python3.11/site-packages/pandas/core/groupby/generic.py", line 1269, in aggregate
    result = op.agg()
             ^^^^^^^^
  File "/home/trevor/miniconda3/envs/pypsa-usa/lib/python3.11/site-packages/pandas/core/apply.py", line 163, in agg
    return self.agg_dict_like()
           ^^^^^^^^^^^^^^^^^^^^
  File "/home/trevor/miniconda3/envs/pypsa-usa/lib/python3.11/site-packages/pandas/core/apply.py", line 420, in agg_dict_like
    results = {
              ^
  File "/home/trevor/miniconda3/envs/pypsa-usa/lib/python3.11/site-packages/pandas/core/apply.py", line 421, in <dictcomp>
    key: obj._gotitem(key, ndim=1).agg(how) for key, how in arg.items()
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/trevor/miniconda3/envs/pypsa-usa/lib/python3.11/site-packages/pandas/core/groupby/generic.py", line 269, in aggregate
    return self._python_agg_general(func, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/trevor/miniconda3/envs/pypsa-usa/lib/python3.11/site-packages/pandas/core/groupby/generic.py", line 288, in _python_agg_general
    result = self.grouper.agg_series(obj, f)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/trevor/miniconda3/envs/pypsa-usa/lib/python3.11/site-packages/pandas/core/groupby/ops.py", line 994, in agg_series
    result = self._aggregate_series_pure_python(obj, func)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/trevor/miniconda3/envs/pypsa-usa/lib/python3.11/site-packages/pandas/core/groupby/ops.py", line 1015, in _aggregate_series_pure_python
    res = func(group)
          ^^^^^^^^^^^
  File "/home/trevor/miniconda3/envs/pypsa-usa/lib/python3.11/site-packages/pandas/core/groupby/generic.py", line 285, in <lambda>
    f = lambda x: func(x, *args, **kwargs)
                  ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/trevor/miniconda3/envs/pypsa-usa/lib/python3.11/site-packages/pypsa/clustering/spatial.py", line 136, in consense
    assert (x == v).all() or x.isnull().all(), (
AssertionError: In Bus cluster LAF_states, the values of attribute LAF_states do not agree:
Bus
2010004    0.000000
2010005    0.000651
Name: LAF_states, dtype: float64
[Sat Feb 24 16:31:22 2024]
Error in rule simplify_network:
    jobid: 7
    input: data/breakthrough_network/base_grid/western/bus2sub.csv, data/breakthrough_network/base_grid/western/sub.csv, resources/western/elec_base_network_l_pp.nc
    output: resources/western/elec_s.nc
    log: logs/simplify_network/western/elec_s.log (check log file(s) for error details)

RuleException:
CalledProcessError in file /home/trevor/repos/pypsa-usa/workflow/rules/build_electricity.smk, line 300:
Command 'set -euo pipefail;  /home/trevor/miniconda3/envs/pypsa-usa/bin/python3.11 /home/trevor/repos/pypsa-usa/workflow/.snakemake/scripts/tmp9zo8mvsl.simplify_network.py' returned non-zero exit status 1.
  File "/home/trevor/repos/pypsa-usa/workflow/rules/build_electricity.smk", line 300, in __rule_simplify_network
  File "/home/trevor/miniconda3/envs/pypsa-usa/lib/python3.11/concurrent/futures/thread.py", line 58, in run
Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
```


### Anything else?

_No response_

--- a/workflow/rules/build_electricity.smk
+++ b/workflow/rules/build_electricity.smk
@@ -210,6 +210,36 @@ rule build_renewable_profiles:
     script:
         "../scripts/build_renewable_profiles.py"
 
+rule build_demand:
+    params:
+        planning_horizons=config["scenario"]["planning_horizons"],
+        snapshots=config["snapshots"]
+    input:
+        base_network=RESOURCES + "{interconnect}/elec_base_network.nc",
+        ads_renewables=(
+            DATA + "WECC_ADS/processed/"
+            if config["network_configuration"] == "ads2032"
+            else []
+        ),
+        ads_2032=(
+            DATA + "WECC_ADS/downloads/2032/Public Data/Hourly Profiles in CSV format"
+            if config["network_configuration"] == "ads2032"
+            else []
+        ),
+        eia=expand(DATA + "GridEmissions/{file}", file=DATAFILES_DMD),
+        efs=DATA + "nrel_efs/EFSLoadProfile_Reference_Moderate.csv",
+    output:
+        demand = RESOURCES + "{interconnect}/demand.csv",
+    log:
+        LOGS + "{interconnect}/build_demand.log",
+    benchmark:
+        BENCHMARKS + "{interconnect}/build_demand"
+    threads: 1
+    resources:
+        mem_mb=10000,
+    script:
+        "../scripts/build_demand.py"
+
 
 rule add_electricity:
     params:
@@ -259,14 +289,7 @@ rule add_electricity:
             if config["network_configuration"] == "ads2032"
             else []
         ),
-        ads_2030=(
-            DATA
-            + "WECC_ADS/downloads/2030/WECC 2030 ADS PCM 2020-12-16 (V1.5) Public Data/CSV Shape Files"
-            if config["network_configuration"] == "ads2032"
-            else []
-        ),
-        eia=expand(DATA + "GridEmissions/{file}", file=DATAFILES_DMD),
-        efs=DATA + "nrel_efs/EFSLoadProfile_Reference_Moderate.csv",
+        demand = RESOURCES + "{interconnect}/demand.csv",
         ng_electric_power_price=DATA + "costs/ng_electric_power_price.csv",
     output:
         RESOURCES + "{interconnect}/elec_base_network_l_pp.nc",

--- a/workflow/rules/build_electricity.smk
+++ b/workflow/rules/build_electricity.smk
@@ -210,10 +210,11 @@ rule build_renewable_profiles:
     script:
         "../scripts/build_renewable_profiles.py"
 
+
 rule build_demand:
     params:
         planning_horizons=config["scenario"]["planning_horizons"],
-        snapshots=config["snapshots"]
+        snapshots=config["snapshots"],
     input:
         base_network=RESOURCES + "{interconnect}/elec_base_network.nc",
         ads_renewables=(
@@ -229,7 +230,7 @@ rule build_demand:
         eia=expand(DATA + "GridEmissions/{file}", file=DATAFILES_DMD),
         efs=DATA + "nrel_efs/EFSLoadProfile_Reference_Moderate.csv",
     output:
-        demand = RESOURCES + "{interconnect}/demand.csv",
+        demand=RESOURCES + "{interconnect}/demand.csv",
     log:
         LOGS + "{interconnect}/build_demand.log",
     benchmark:
@@ -289,7 +290,7 @@ rule add_electricity:
             if config["network_configuration"] == "ads2032"
             else []
         ),
-        demand = RESOURCES + "{interconnect}/demand.csv",
+        demand=RESOURCES + "{interconnect}/demand.csv",
         ng_electric_power_price=DATA + "costs/ng_electric_power_price.csv",
     output:
         RESOURCES + "{interconnect}/elec_base_network_l_pp.nc",

--- a/workflow/scripts/add_electricity.py
+++ b/workflow/scripts/add_electricity.py
@@ -1,6 +1,6 @@
 # PyPSA USA Authors
 """
-Adds electrical generators and existing hydro storage units to a base network
+Adds electrical generators, storage units, and demand to a base network
 based on the given network configuration.
 
 **Relevant Settings**
@@ -44,25 +44,15 @@ based on the given network configuration.
 **Inputs**
 
 - ``resources/costs.csv``: The database of cost assumptions for all included technologies for specific years from various sources; e.g. discount rate, lifetime, investment (CAPEX), fixed operation and maintenance (FOM), variable operation and maintenance (VOM), fuel costs, efficiency, carbon-dioxide intensity.
-- ``data/bundle/hydro_capacities.csv``: Hydropower plant store/discharge power capacities, energy storage capacity, and average hourly inflow by country.
-
-    # .. image:: _static/hydrocapacities.png
-    #     :scale: 34 %
-
-- ``data/geth2015_hydro_capacities.csv``: alternative to capacities above; not currently used!
-- ``resources/load.csv`` Hourly per-country load profiles.
+- ``resources/demand.csv`` Hourly per-country load profiles.
 - ``resources/regions_onshore.geojson``: confer :ref:`busregions`
-- ``resources/nuts3_shapes.geojson``: confer :ref:`shapes`
 - ``resources/profile_{}.nc``: all technologies in ``config["renewables"].keys()``, confer :ref:`renewableprofiles`.
-- ``networks/base.nc``: confer :ref:`base`
+- ``networks/elec_base_network.nc``: confer :ref:`base`
 
 **Outputs**
 
+- ``networks/elec_base_network_l_pp.nc``
 
-- ``networks/elec.nc``:
-
-    # .. image:: _static/elec.png
-    #         :scale: 33 %
 
 **Description**
 """
@@ -771,175 +761,14 @@ def attach_renewable_capacities_to_atlite(n, plants_df, renewable_carriers):
         n.generators.p_nom_min.update(generators_tech.bus.map(caps_per_bus).dropna())
 
 
-def prepare_ads_demand(
-    n: pypsa.Network,
-    demand_path: str,
-) -> pd.DataFrame:
-    demand = pd.read_csv(demand_path, index_col=0)
-    data_year = 2032
-    demand.columns = demand.columns.str.removeprefix("Load_")
-    demand.columns = demand.columns.str.removesuffix(".dat")
-    demand.columns = demand.columns.str.removesuffix(f"_{data_year}")
-    demand.columns = demand.columns.str.removesuffix(f"_[18].dat: {data_year}")
-    demand["CISO-PGAE"] = (
-        demand.pop("CIPV") + demand.pop("CIPB") + demand.pop("SPPC")
-    )  # TODO: #37 Create new Zone for SPPC
-    demand["BPAT"] = demand.pop("BPAT") + demand.pop("TPWR") + demand.pop("SCL")
-    demand["IPCO"] = demand.pop("IPFE") + demand.pop("IPMV") + demand.pop("IPTV")
-    demand["PACW"] = demand.pop("PAID") + demand.pop("PAUT") + demand.pop("PAWY")
-    demand["Arizona"] = demand.pop("SRP") + demand.pop("AZPS")
-    demand.drop(columns=["Unnamed: 44", "TH_Malin", "TH_Mead", "TH_PV"], inplace=True)
-    ba_list_map = {
-        "CISC": "CISO-SCE",
-        "CISD": "CISO-SDGE",
-        "VEA": "CISO-VEA",
-        "WAUW": "WAUW_SPP",
-    }
-    demand.rename(columns=ba_list_map, inplace=True)
-    demand.set_index("Index", inplace=True)
-
-    demand.index = n.snapshots
-    n.buses["load_dissag"] = n.buses.balancing_area.replace({"": "missing_ba"})
-    intersection = set(demand.columns).intersection(n.buses.ba_load_data.unique())
-    demand = demand[list(intersection)]
-    set_load_allocation_factor(n)
-    return disaggregate_demand_to_buses(n, demand)
-
-
-def prepare_eia_demand(
-    n: pypsa.Network,
-    demand_path: str,
-) -> pd.DataFrame:
-    logger.info("Building Load Data using EIA demand")
-    demand = pd.read_csv(demand_path, index_col=0)
-    demand.index = pd.to_datetime(demand.index)
-    demand = demand.loc[n.snapshots.intersection(demand.index)]  # filter by snapshots
-    demand = demand[~demand.index.duplicated(keep="last")]
-    demand.index = n.snapshots
-
-    # Combine EIA Demand Data to Match GIS Shapes
-    demand["Arizona"] = demand.pop("SRP") + demand.pop("AZPS")
-    n.buses["load_dissag"] = n.buses.balancing_area.replace(
-        {"^CISO.*": "CISO", "^ERCO.*": "ERCO"},
-        regex=True,
-    )
-
-    n.buses["load_dissag"] = n.buses.load_dissag.replace({"": "missing_ba"})
-    intersection = set(demand.columns).intersection(n.buses.load_dissag.unique())
-    demand = demand[list(intersection)]
-
-    set_load_allocation_factor(n)
-    return disaggregate_demand_to_buses(n, demand)
-
-
-def prepare_efs_demand(
-    n: pypsa.Network,
-    planning_horizons: list[str],
-) -> pd.DataFrame:
-    logger.info("Building Load Data using EFS demand")
-    demand = pd.read_csv(snakemake.input.efs)
-
-    demand = demand.loc[demand.Year == planning_horizons[0]]
-    demand.drop(
-        columns=[
-            "Electrification",
-            "TechnologyAdvancement",
-            "Sector",
-            "Subsector",
-            "Year",
-        ],
-        inplace=True,
-    )
-    # TODO: We are throwing out great data here on the sector and subsector loads. Revisit this.
-
-    demand = demand.groupby(["State", "LocalHourID"]).sum().reset_index()
-
-    demand["DateTime"] = pd.Timestamp(
-        year=planning_horizons[0],
-        month=1,
-        day=1,
-    ) + pd.to_timedelta(demand["LocalHourID"] - 1, unit="H")
-    demand["UTC_Time"] = demand.groupby(["State"])["DateTime"].transform(local_to_utc)
-    demand.drop(columns=["LocalHourID", "DateTime"], inplace=True)
-    demand.set_index("UTC_Time", inplace=True)
-
-    demand = demand.pivot(columns="State", values="LoadMW")
-    n.buses["load_dissag"] = n.buses.state
-    intersection = set(demand.columns).intersection(
-        [const.STATE_2_CODE.get(item, item) for item in n.buses.load_dissag.unique()],
-    )
-    demand = demand[list(intersection)]
-    demand.columns = [
-        {v: k for k, v in const.STATE_2_CODE.items()}.get(item, item)
-        for item in demand.columns
-    ]
-    demand = demand.dropna()
-    if snakemake.wildcards.interconnect == "texas":
-        demand = demand.iloc[2:, :]  # temp fix for lining up timezones
-    year = n.snapshots[0].year
-    demand.index = demand.index.map(lambda x: x.replace(year=year))
-    demand = demand.loc[n.snapshots.intersection(demand.index)]
-    n.buses.rename(columns={"LAF_states": "LAF"}, inplace=True)
-    # demand.loc[:,'Texas'] = demand.loc[:,'Texas'] / 400 #temp
-    return disaggregate_demand_to_buses(n, demand)
-
-
-def set_load_allocation_factor(n: pypsa.Network) -> pd.DataFrame:
-    """
-    Defines Load allocation factor for each bus according to load_dissag for
-    balancing areas.
-    """
-    n.buses.Pd = n.buses.Pd.fillna(0)
-    group_sums = n.buses.groupby("load_dissag")["Pd"].transform("sum")
-    n.buses["LAF"] = n.buses["Pd"] / group_sums
-
-
-def disaggregate_demand_to_buses(
-    n: pypsa.Network,
-    demand: pd.DataFrame,
-) -> pd.DataFrame:
-    """
-    Zone power demand is disaggregated to buses proportional to Pd.
-    """
-    demand_aligned = demand.reindex(
-        columns=n.buses["load_dissag"].unique(),
-        fill_value=0,
-    )
-    bus_demand = pd.DataFrame()
-    for load_dissag in n.buses["load_dissag"].unique():
-        LAF = n.buses.loc[n.buses["load_dissag"] == load_dissag, "LAF"]
-        zone_bus_demand = (
-            demand_aligned[load_dissag].values.reshape(-1, 1) * LAF.values.T
-        )
-        bus_demand = pd.concat(
-            [bus_demand, pd.DataFrame(zone_bus_demand, columns=LAF.index)],
-            axis=1,
-        )
-    bus_demand.index = n.snapshots
-    n.buses.drop(columns=["LAF"], inplace=True)
-    return bus_demand.fillna(0)
-
-
-def attach_demand(n: pypsa.Network, configuration: str):
+def attach_demand(n: pypsa.Network, demand_per_bus_fn: str):
     """
     Add demand to network from specified configuration setting.
 
     Returns network with demand added.
     """
-    if configuration == "ads":
-        demand_per_bus = prepare_ads_demand(n, f"data/WECC_ADS/processed/load_2032.csv")
-    elif configuration == "pypsa-usa":
-        if snakemake.params.get("planning_horizons"):
-            demand_per_bus = prepare_efs_demand(
-                n,
-                snakemake.params.get("planning_horizons"),
-            )
-        else:
-            demand_per_bus = prepare_eia_demand(n, snakemake.input["eia"][0])
-    else:
-        raise ValueError(
-            "Invalid demand_type. Supported values are 'ads', and 'pypsa-usa'.",
-        )
+    demand_per_bus = pd.read_csv(demand_per_bus_fn, index_col=0)
+    demand_per_bus.index = pd.to_datetime(demand_per_bus.index)
     n.madd(
         "Load",
         demand_per_bus.columns,
@@ -1443,8 +1272,11 @@ def load_powerplants_ads(
 
 
 def clean_bus_data(n: pypsa.Network):
-    col_list = ["poi_bus", "poi_sub", "poi", "Pd", "load_dissag"]
-    n.buses.drop(columns=col_list, inplace=True)
+    """
+    Drops data from the network that are no longer needed in workflow.
+    """
+    col_list = ["poi_bus", "poi_sub", "poi", "Pd", "load_dissag", "LAF","LAF_states"]
+    n.buses.drop(columns=[col for col in col_list if col in n.buses], inplace=True)
 
 
 def main(snakemake):
@@ -1456,8 +1288,8 @@ def main(snakemake):
     n = pypsa.Network(snakemake.input.base_network)
 
     snapshot_config = snakemake.config["snapshots"]
-    sns_start = pd.to_datetime(snapshot_config["start"] + " 08:00:00")
-    sns_end = pd.to_datetime(snapshot_config["end"] + " 06:00:00")
+    sns_start = pd.to_datetime(snapshot_config["start"])
+    sns_end = pd.to_datetime(snapshot_config["end"])
     sns_inclusive = snapshot_config["inclusive"]
 
     n.set_snapshots(
@@ -1533,7 +1365,8 @@ def main(snakemake):
 
     # Applying to all configurations
     plants = match_plant_to_bus(n, plants)
-    attach_demand(n, configuration)
+    
+    attach_demand(n, snakemake.input.demand)
 
     attach_conventional_generators(
         n,

--- a/workflow/scripts/add_electricity.py
+++ b/workflow/scripts/add_electricity.py
@@ -1,7 +1,6 @@
 # PyPSA USA Authors
 """
-Adds electrical generators, storage units, and demand to a base network
-based on the given network configuration.
+Add_electricity takes data produced by build_renewable_profiles, build_demand, build_cost_data and build_base_network to create a combined network model of all the generators, demand, costs. Locational multipliers are added for regional fuel costs and capital costs. 
 
 **Relevant Settings**
 

--- a/workflow/scripts/add_electricity.py
+++ b/workflow/scripts/add_electricity.py
@@ -1,6 +1,9 @@
 # PyPSA USA Authors
 """
-Add_electricity takes data produced by build_renewable_profiles, build_demand, build_cost_data and build_base_network to create a combined network model of all the generators, demand, costs. Locational multipliers are added for regional fuel costs and capital costs. 
+Add_electricity takes data produced by build_renewable_profiles, build_demand,
+build_cost_data and build_base_network to create a combined network model of
+all the generators, demand, costs. Locational multipliers are added for
+regional fuel costs and capital costs.
 
 **Relevant Settings**
 
@@ -1274,7 +1277,7 @@ def clean_bus_data(n: pypsa.Network):
     """
     Drops data from the network that are no longer needed in workflow.
     """
-    col_list = ["poi_bus", "poi_sub", "poi", "Pd", "load_dissag", "LAF","LAF_states"]
+    col_list = ["poi_bus", "poi_sub", "poi", "Pd", "load_dissag", "LAF", "LAF_states"]
     n.buses.drop(columns=[col for col in col_list if col in n.buses], inplace=True)
 
 
@@ -1364,7 +1367,7 @@ def main(snakemake):
 
     # Applying to all configurations
     plants = match_plant_to_bus(n, plants)
-    
+
     attach_demand(n, snakemake.input.demand)
 
     attach_conventional_generators(

--- a/workflow/scripts/build_demand.py
+++ b/workflow/scripts/build_demand.py
@@ -1,6 +1,9 @@
 # PyPSA USA Authors
 """
-Builds the demand data for the PyPSA network. This module is responsible for cleaning and transforming electricity demand data from the NREL Electrification Futures Study, EIA, and GridEmissions to be used in the `add_electricity` module. 
+Builds the demand data for the PyPSA network. This module is responsible for
+cleaning and transforming electricity demand data from the NREL Electrification
+Futures Study, EIA, and GridEmissions to be used in the `add_electricity`
+module.
 
 **Relevant Settings**
 
@@ -20,7 +23,7 @@ Builds the demand data for the PyPSA network. This module is responsible for cle
 
 **Inputs**
 
-    - base_network:  
+    - base_network:
     - ads_renewables:
     - ads_2032:
     - eia: (GridEmissions data file)
@@ -123,6 +126,7 @@ def prepare_eia_demand(
     set_load_allocation_factor(n)
     return disaggregate_demand_to_buses(n, demand)
 
+
 def prepare_efs_demand(
     n: pypsa.Network,
     planning_horizons: list[str],
@@ -165,13 +169,21 @@ def prepare_efs_demand(
         for item in demand.columns
     ]
 
-    #This block is use to align the demand data hours with the snapshot hours
+    # This block is use to align the demand data hours with the snapshot hours
     hoy = (demand.index.dayofyear - 1) * 24 + demand.index.hour
     demand.index = hoy
     demand_new = pd.DataFrame(columns=demand.columns)
     for column in demand.columns:
         col = demand[column].reset_index()
-        demand_new[column] = col.groupby('UTC_Time').apply(lambda group: group.loc[group.drop(columns='UTC_Time').first_valid_index()]).drop(columns='UTC_Time')
+        demand_new[column] = (
+            col.groupby("UTC_Time")
+            .apply(
+                lambda group: group.loc[
+                    group.drop(columns="UTC_Time").first_valid_index()
+                ]
+            )
+            .drop(columns="UTC_Time")
+        )
 
     demand_new.index = n.snapshots
     n.buses.rename(columns={"LAF_states": "LAF"}, inplace=True)
@@ -220,10 +232,9 @@ def main(snakemake):
     interconnection = snakemake.wildcards["interconnect"]
     planning_horizons = snakemake.params["planning_horizons"]
 
-
     snapshot_config = snakemake.params["snapshots"]
-    sns_start = pd.to_datetime(snapshot_config["start"]) #+ " 08:00:00")
-    sns_end = pd.to_datetime(snapshot_config["end"]) #+ " 06:00:00")
+    sns_start = pd.to_datetime(snapshot_config["start"])  # + " 08:00:00")
+    sns_end = pd.to_datetime(snapshot_config["end"])  # + " 06:00:00")
     sns_inclusive = snapshot_config["inclusive"]
 
     n = pypsa.Network(snakemake.input.base_network)
@@ -241,9 +252,15 @@ def main(snakemake):
     if configuration == "ads":
         demand_per_bus = prepare_ads_demand(n, "data/WECC_ADS/processed/load_2032.csv")
     elif configuration == "pypsa-usa":
-        demand_per_bus = prepare_efs_demand(n, snakemake.params.get("planning_horizons")) if snakemake.params.get("planning_horizons") else prepare_eia_demand(n, snakemake.input["eia"][0])
+        demand_per_bus = (
+            prepare_efs_demand(n, snakemake.params.get("planning_horizons"))
+            if snakemake.params.get("planning_horizons")
+            else prepare_eia_demand(n, snakemake.input["eia"][0])
+        )
     else:
-        raise ValueError("Invalid demand_type. Supported values are 'ads', and 'pypsa-usa'.")
+        raise ValueError(
+            "Invalid demand_type. Supported values are 'ads', and 'pypsa-usa'."
+        )
 
     demand_per_bus.to_csv(snakemake.output.demand, index=True)
 

--- a/workflow/scripts/build_demand.py
+++ b/workflow/scripts/build_demand.py
@@ -1,0 +1,254 @@
+# PyPSA USA Authors
+"""
+Builds the demand data for the PyPSA network. This module is responsible for cleaning and transforming electricity demand data from the NREL Electrification Futures Study, EIA, and GridEmissions to be used in the `add_electricity` module. 
+
+**Relevant Settings**
+
+.. code:: yaml
+
+    network_configuration:
+
+    snapshots:
+        start:
+        end:
+        inclusive:
+
+    scenario:
+    interconnect:
+    planning_horizons:
+
+
+**Inputs**
+
+
+
+**Outputs**
+
+
+
+"""
+
+
+import logging
+import os
+import random
+from itertools import product
+from pathlib import Path
+from typing import Any
+from typing import Dict
+from typing import List
+from typing import Union
+
+import constants as const
+import geopandas as gpd
+import numpy as np
+import pandas as pd
+import pypsa
+import xarray as xr
+from _helpers import configure_logging
+from _helpers import export_network_for_gis_mapping
+from _helpers import local_to_utc
+from _helpers import test_network_datatype_consistency
+from _helpers import update_p_nom_max
+from scipy import sparse
+from shapely.geometry import Point
+from shapely.prepared import prep
+from sklearn.neighbors import BallTree
+
+idx = pd.IndexSlice
+
+logger = logging.getLogger(__name__)
+
+
+def prepare_ads_demand(
+    n: pypsa.Network,
+    demand_path: str,
+) -> pd.DataFrame:
+    demand = pd.read_csv(demand_path, index_col=0)
+    data_year = 2032
+    demand.columns = demand.columns.str.removeprefix("Load_")
+    demand.columns = demand.columns.str.removesuffix(".dat")
+    demand.columns = demand.columns.str.removesuffix(f"_{data_year}")
+    demand.columns = demand.columns.str.removesuffix(f"_[18].dat: {data_year}")
+    demand["CISO-PGAE"] = (
+        demand.pop("CIPV") + demand.pop("CIPB") + demand.pop("SPPC")
+    )  # TODO: #37 Create new Zone for SPPC
+    demand["BPAT"] = demand.pop("BPAT") + demand.pop("TPWR") + demand.pop("SCL")
+    demand["IPCO"] = demand.pop("IPFE") + demand.pop("IPMV") + demand.pop("IPTV")
+    demand["PACW"] = demand.pop("PAID") + demand.pop("PAUT") + demand.pop("PAWY")
+    demand["Arizona"] = demand.pop("SRP") + demand.pop("AZPS")
+    demand.drop(columns=["Unnamed: 44", "TH_Malin", "TH_Mead", "TH_PV"], inplace=True)
+    ba_list_map = {
+        "CISC": "CISO-SCE",
+        "CISD": "CISO-SDGE",
+        "VEA": "CISO-VEA",
+        "WAUW": "WAUW_SPP",
+    }
+    demand.rename(columns=ba_list_map, inplace=True)
+    demand.set_index("Index", inplace=True)
+
+    demand.index = n.snapshots
+    n.buses["load_dissag"] = n.buses.balancing_area.replace({"": "missing_ba"})
+    intersection = set(demand.columns).intersection(n.buses.ba_load_data.unique())
+    demand = demand[list(intersection)]
+    set_load_allocation_factor(n)
+    return disaggregate_demand_to_buses(n, demand)
+
+
+def prepare_eia_demand(
+    n: pypsa.Network,
+    demand_path: str,
+) -> pd.DataFrame:
+    logger.info("Building Load Data using EIA demand")
+    demand = pd.read_csv(demand_path, index_col=0)
+    demand.index = pd.to_datetime(demand.index)
+    demand = demand.loc[n.snapshots.intersection(demand.index)]  # filter by snapshots
+    demand = demand[~demand.index.duplicated(keep="last")]
+    demand.index = n.snapshots
+
+    # Combine EIA Demand Data to Match GIS Shapes
+    demand["Arizona"] = demand.pop("SRP") + demand.pop("AZPS")
+    n.buses["load_dissag"] = n.buses.balancing_area.replace(
+        {"^CISO.*": "CISO", "^ERCO.*": "ERCO"},
+        regex=True,
+    )
+
+    n.buses["load_dissag"] = n.buses.load_dissag.replace({"": "missing_ba"})
+    intersection = set(demand.columns).intersection(n.buses.load_dissag.unique())
+    demand = demand[list(intersection)]
+
+    set_load_allocation_factor(n)
+    return disaggregate_demand_to_buses(n, demand)
+
+def prepare_efs_demand(
+    n: pypsa.Network,
+    planning_horizons: list[str],
+) -> pd.DataFrame:
+    logger.info("Building Load Data using EFS demand")
+
+    demand = pd.read_csv(snakemake.input.efs)
+    demand = demand.loc[demand.Year == planning_horizons[0]]
+    demand.drop(
+        columns=[
+            "Electrification",
+            "TechnologyAdvancement",
+            "Sector",
+            "Subsector",
+            "Year",
+        ],
+        inplace=True,
+    )
+    # TODO: We are throwing out great data here on the sector and subsector loads. Revisit this.
+
+    demand = demand.groupby(["State", "LocalHourID"]).sum().reset_index()
+
+    demand["DateTime"] = pd.Timestamp(
+        year=planning_horizons[0],
+        month=1,
+        day=1,
+    ) + pd.to_timedelta(demand["LocalHourID"] - 1, unit="H")
+    demand["UTC_Time"] = demand.groupby(["State"])["DateTime"].transform(local_to_utc)
+    demand.drop(columns=["LocalHourID", "DateTime"], inplace=True)
+    demand.set_index("UTC_Time", inplace=True)
+
+    demand = demand.pivot(columns="State", values="LoadMW")
+    n.buses["load_dissag"] = n.buses.state
+    intersection = set(demand.columns).intersection(
+        [const.STATE_2_CODE.get(item, item) for item in n.buses.load_dissag.unique()],
+    )
+    demand = demand[list(intersection)]
+    demand.columns = [
+        {v: k for k, v in const.STATE_2_CODE.items()}.get(item, item)
+        for item in demand.columns
+    ]
+
+    #This block is use to align the demand data hours with the snapshot hours
+    hoy = (demand.index.dayofyear - 1) * 24 + demand.index.hour
+    demand.index = hoy
+    demand_new = pd.DataFrame(columns=demand.columns)
+    for column in demand.columns:
+        col = demand[column].reset_index()
+        demand_new[column] = col.groupby('UTC_Time').apply(lambda group: group.loc[group.drop(columns='UTC_Time').first_valid_index()]).drop(columns='UTC_Time')
+
+    demand_new.index = n.snapshots
+    n.buses.rename(columns={"LAF_states": "LAF"}, inplace=True)
+    return disaggregate_demand_to_buses(n, demand_new)
+
+
+def set_load_allocation_factor(n: pypsa.Network) -> pd.DataFrame:
+    """
+    Defines Load allocation factor for each bus according to load_dissag for
+    balancing areas.
+    """
+    n.buses.Pd = n.buses.Pd.fillna(0)
+    group_sums = n.buses.groupby("load_dissag")["Pd"].transform("sum")
+    n.buses["LAF"] = n.buses["Pd"] / group_sums
+
+
+def disaggregate_demand_to_buses(
+    n: pypsa.Network,
+    demand: pd.DataFrame,
+) -> pd.DataFrame:
+    """
+    Zone power demand is disaggregated to buses proportional to Pd.
+    """
+    demand_aligned = demand.reindex(
+        columns=n.buses["load_dissag"].unique(),
+        fill_value=0,
+    )
+    bus_demand = pd.DataFrame()
+    for load_dissag in n.buses["load_dissag"].unique():
+        LAF = n.buses.loc[n.buses["load_dissag"] == load_dissag, "LAF"]
+        zone_bus_demand = (
+            demand_aligned[load_dissag].values.reshape(-1, 1) * LAF.values.T
+        )
+        bus_demand = pd.concat(
+            [bus_demand, pd.DataFrame(zone_bus_demand, columns=LAF.index)],
+            axis=1,
+        )
+    bus_demand.index = n.snapshots
+    n.buses.drop(columns=["LAF"], inplace=True)
+    return bus_demand.fillna(0)
+
+
+def main(snakemake):
+    params = snakemake.params
+    configuration = snakemake.config["network_configuration"]
+    interconnection = snakemake.wildcards["interconnect"]
+    planning_horizons = snakemake.params["planning_horizons"]
+
+
+    snapshot_config = snakemake.params["snapshots"]
+    sns_start = pd.to_datetime(snapshot_config["start"]) #+ " 08:00:00")
+    sns_end = pd.to_datetime(snapshot_config["end"]) #+ " 06:00:00")
+    sns_inclusive = snapshot_config["inclusive"]
+
+    n = pypsa.Network(snakemake.input.base_network)
+
+    n.set_snapshots(
+        pd.date_range(
+            freq="h",
+            start=sns_start,
+            end=sns_end,
+            inclusive=sns_inclusive,
+        ),
+    )
+    Nyears = n.snapshot_weightings.objective.sum() / 8760.0
+
+    if configuration == "ads":
+        demand_per_bus = prepare_ads_demand(n, "data/WECC_ADS/processed/load_2032.csv")
+    elif configuration == "pypsa-usa":
+        demand_per_bus = prepare_efs_demand(n, snakemake.params.get("planning_horizons")) if snakemake.params.get("planning_horizons") else prepare_eia_demand(n, snakemake.input["eia"][0])
+    else:
+        raise ValueError("Invalid demand_type. Supported values are 'ads', and 'pypsa-usa'.")
+
+    demand_per_bus.to_csv(snakemake.output.demand, index=True)
+
+
+if __name__ == "__main__":
+    if "snakemake" not in globals():
+        from _helpers import mock_snakemake
+
+        snakemake = mock_snakemake("build_demand", interconnect="western")
+    configure_logging(snakemake)
+    main(snakemake)

--- a/workflow/scripts/build_demand.py
+++ b/workflow/scripts/build_demand.py
@@ -20,12 +20,15 @@ Builds the demand data for the PyPSA network. This module is responsible for cle
 
 **Inputs**
 
-
+    - base_network:  
+    - ads_renewables:
+    - ads_2032:
+    - eia: (GridEmissions data file)
+    - efs: (NREL EFS Load Forecasts)
 
 **Outputs**
 
-
-
+    - demand: Path to the demand CSV file.
 """
 
 

